drivers: sensor: add it51xxx_evb adc voltage comparator driver

Yunshao-Chiang opened the pull request:
Add voltage comparator driver for ITE it51xxx chip.

Yunshao-Chiang committed:
drivers: comparator: add it51xxx_evb analog comparator driver

Add analog comparator driver for ITE it51xxx chip.

Signed-off-by: Yunshao Chiang <Yunshao.Chiang@ite.com.tw>

Dino-Li commented on boards/ite/it515xx_evb/Kconfig.defconfig:
```diff
@@ -14,4 +14,12 @@ endchoice
+if VCMP_IT51XXX
+
+choice VCMP_IT51XXX_SCAN_PERIOD
+	default VCMP_IT51XXX_SCAN_PERIOD_1MS
```
Can we add new property of node to set scan period?

Dino-Li commented on drivers/sensor/ite/ite_vcmp_it51xxx/Kconfig:
```diff
@@ -0,0 +1,90 @@
+config VCMP_IT51XXX_SCAN_PERIOD_5MS
+	bool "VCMP_IT51XXX_SCAN_PERIOD_5MS"
+
+endchoice
```
If we can use a property to set the scan period, we can drop this choice.

Dino-Li commented on drivers/sensor/ite/ite_vcmp_it51xxx/vcmp_ite_it51xxx.c:
```diff
@@ -0,0 +1,381 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(vcmp_ite_it51xxx, CONFIG_SENSOR_LOG_LEVEL);
+
+#define VCMP_CHANNEL_ID_REG_MASK 0x7
```
Use GENMASK() here.

Dino-Li commented on include/zephyr/dt-bindings/sensor/it51xxx_vcmp.h:
```diff
@@ -0,0 +1,65 @@
+#define VCMP_SCAN_PERIOD 0xB
+#elif CONFIG_VCMP_IT51XXX_SCAN_PERIOD_5MS
+#define VCMP_SCAN_PERIOD 0xC
+#endif
```
Drop these macros If we can use a property to set scan period.

Dino-Li commented on dts/riscv/ite/it51xxx.dtsi:
```diff
@@ -80,6 +81,38 @@
 			#io-channel-cells = <1>;
 		};
 
+		vcmp_base: vcmp_base@f04500 {
```
Is this node necessary? (This reg base can be listed to each vcmp node below)

Dino-Li commented:
@MaureenHelm any comments will be appreciated.

Yunshao-Chiang commented on drivers/sensor/ite/ite_vcmp_it51xxx/Kconfig:
```diff
@@ -0,0 +1,48 @@
+	bool "ITE it51xxx Voltage Comparator"
+	default y
+	depends on DT_HAS_ITE_IT51XXX_VCMP_ENABLED
+	depends on SOC_IT51XXX && ADC_ITE_IT51XXX
```
Yes, it is.
I removed SOC_IT51XXX.

Yunshao-Chiang commented on drivers/sensor/ite/ite_vcmp_it51xxx/vcmp_ite_it51xxx.c:
```diff
@@ -0,0 +1,385 @@
+const uint8_t vcmp_ctrl_reg[3] = {REG_VCMP_VCMP0SCTL, REG_VCMP_VCMP1SCTL, REG_VCMP_VCMP2SCTL};
+BUILD_ASSERT(ARRAY_SIZE(vcmp_ctrl_reg) == VCMP_CHANNEL_CNT,
+	     "The number of elements in `vcmp_ctrl_reg` array must be equal to `VCMP_CHANNEL_CNT`, "
+	     "and all elements must be initialized with their corresponding addresses");
```
It is used to ensure that when someone adds a `VCMP_CHANNEL_X`, they also add the corresponding address of the VCMP control register.

After thinking about it, when modifying the VCMP control register macro, one would probably notice that the array also needs to be updated.

removed it.

Yunshao-Chiang commented on drivers/sensor/ite/ite_vcmp_it51xxx/vcmp_ite_it51xxx.c:
```diff
@@ -0,0 +1,385 @@
+		.adc_ch_cfg.reference = ADC_REF_INTERNAL,                                          \
+		.adc_ch_cfg.acquisition_time = ADC_ACQ_TIME_DEFAULT,                               \
+		.adc_ch_cfg.channel_id = (uint8_t)DT_INST_IO_CHANNELS_INPUT(inst),                 \
+	};                                                                                         \
```
After the code changes, a total of 108 bytes are saved when all three voltage comparators are used, with each comparator contributing a 36-byte reduction

fabiobaltieri replied:
neat!

Yunshao-Chiang commented:
Hi @MaureenHelm,

When you have a moment, could you please take a look at my PR?
I'd really appreciate your feedback.

Dino-Li commented:
@MaureenHelm any comments will be appreciated.

MaureenHelm reviewed (changes requested):
This should implement the comparator API, not the sensor API. See #76087

Yunshao-Chiang commented:
Hi @MaureenHelm,

Since it51xxx does not support `COMPARATOR_TRIGGER_FALLING_EDGE`, `COMPARATOR_TRIGGER_RISING_EDGE`, and `COMPARATOR_TRIGGER_BOTH_EDGES`, could I add something like `COMPARATOR_TRIGGER_LEVEL_SENSED_MODE` in [comparator.h](https://github.com/zephyrproject-rtos/zephyr/blob/b5934e8dfe5c8dd4c1a9f2221341aa46e3dce72a/include/zephyr/drivers/comparator.h#L27-L36) as follows?

Level-sensed mode triggers when the trigger condition is met, and it continues to trigger as long as the condition remains true, unlike rising edge or falling edge modes which only trigger at the instant of the upward or downward transition.

```diff
diff --git a/include/zephyr/drivers/comparator.h b/include/zephyr/drivers/comparator.h
index 1a6609cafa5..d7a35bf9426 100644
--- a/include/zephyr/drivers/comparator.h
+++ b/include/zephyr/drivers/comparator.h
@@ -32,7 +32,9 @@ enum comparator_trigger {
        /** Trigger on falling edge of comparator output */
        COMPARATOR_TRIGGER_FALLING_EDGE,
        /** Trigger on both edges of comparator output */
-       COMPARATOR_TRIGGER_BOTH_EDGES
+       COMPARATOR_TRIGGER_BOTH_EDGES,
+       /** Trigger on level-sensed mode of comparator output */
+       COMPARATOR_TRIGGER_LEVEL_SENSED_MODE
 };
 
 /** Comparator callback template */
```

MaureenHelm commented:
> Hi @MaureenHelm,
> 
> Since it51xxx does not support `COMPARATOR_TRIGGER_FALLING_EDGE`, `COMPARATOR_TRIGGER_RISING_EDGE`, and `COMPARATOR_TRIGGER_BOTH_EDGES`, could I add something like `COMPARATOR_TRIGGER_LEVEL_SENSED_MODE` in [comparator.h](https://github.com/zephyrproject-rtos/zephyr/blob/b5934e8dfe5c8dd4c1a9f2221341aa46e3dce72a/include/zephyr/drivers/comparator.h#L27-L36) as follows?
> 
> Level-sensed mode triggers when the trigger condition is met, and it continues to trigger as long as the condition remains true, unlike rising edge or falling edge modes which only trigger at the instant of the upward or downward transition.
> 
> ```diff
> diff --git a/include/zephyr/drivers/comparator.h b/include/zephyr/drivers/comparator.h
> index 1a6609cafa5..d7a35bf9426 100644
> --- a/include/zephyr/drivers/comparator.h
> +++ b/include/zephyr/drivers/comparator.h
> @@ -32,7 +32,9 @@ enum comparator_trigger {
>         /** Trigger on falling edge of comparator output */
>         COMPARATOR_TRIGGER_FALLING_EDGE,
>         /** Trigger on both edges of comparator output */
> -       COMPARATOR_TRIGGER_BOTH_EDGES
> +       COMPARATOR_TRIGGER_BOTH_EDGES,
> +       /** Trigger on level-sensed mode of comparator output */
> +       COMPARATOR_TRIGGER_LEVEL_SENSED_MODE
>  };
>  
>  /** Comparator callback template */
> ```

^^ @bjarki-andreasen

bjarki-andreasen reviewed (changes requested):
Comparators shall use the comparator API, see https://docs.zephyrproject.org/latest/hardware/peripherals/comparator.html

bjarki-andreasen commented:
> > Hi @MaureenHelm,
> > Since it51xxx does not support `COMPARATOR_TRIGGER_FALLING_EDGE`, `COMPARATOR_TRIGGER_RISING_EDGE`, and `COMPARATOR_TRIGGER_BOTH_EDGES`, could I add something like `COMPARATOR_TRIGGER_LEVEL_SENSED_MODE` in [comparator.h](https://github.com/zephyrproject-rtos/zephyr/blob/b5934e8dfe5c8dd4c1a9f2221341aa46e3dce72a/include/zephyr/drivers/comparator.h#L27-L36) as follows?
> > Level-sensed mode triggers when the trigger condition is met, and it continues to trigger as long as the condition remains true, unlike rising edge or falling edge modes which only trigger at the instant of the upward or downward transition.
> > ```diff
> > diff --git a/include/zephyr/drivers/comparator.h b/include/zephyr/drivers/comparator.h
> > index 1a6609cafa5..d7a35bf9426 100644
> > --- a/include/zephyr/drivers/comparator.h
> > +++ b/include/zephyr/drivers/comparator.h
> > @@ -32,7 +32,9 @@ enum comparator_trigger {
> >         /** Trigger on falling edge of comparator output */
> >         COMPARATOR_TRIGGER_FALLING_EDGE,
> >         /** Trigger on both edges of comparator output */
> > -       COMPARATOR_TRIGGER_BOTH_EDGES
> > +       COMPARATOR_TRIGGER_BOTH_EDGES,
> > +       /** Trigger on level-sensed mode of comparator output */
> > +       COMPARATOR_TRIGGER_LEVEL_SENSED_MODE
> >  };
> >  
> >  /** Comparator callback template */
> > ```
> 
> ^^ @bjarki-andreasen

Hi, though the interrupt may be level triggered, this is not useful to the user, and it is up to the driver to convert this to a single event. This is typically done by enabling the level triggered interrupt in either high or low level (since level can't be both), wait for the level to be reached (this is the rising or falling edge defined by the comparator API), driver reacts to the interrupt, driver masks the level triggered interrupt, calls the user callback, enables opposite level triggered interrupt. It then bounces between these, calling the user callback when the edge matches what they have requested through the comparator_set_trigger API :)

Yunshao-Chiang commented:
Hi @MaureenHelm,

Apologies for the long delay. I've refactored the voltage comparator to use the Comparator API instead of the Sensor API.
When you have a moment, I'd really appreciate your review on this PR. Thanks in advance for your time!

Dino-Li commented:
@MaureenHelm @bjarki-andreasen please revisit.

bjarki-andreasen commented on drivers/comparator/Kconfig.it51xxx_vcmp:
```diff
@@ -0,0 +1,48 @@
+	help
+	  This option sets ITE voltage comparator device instance init priority.
+
+config VCMP_IT51XXX_WORKQUEUE
+	bool "ITE it51xxx voltage comparator threshold detection uses internal work queue"
+	help
+	  Threshold detection ISR utilizes system work queue for calling
+	  trigger handlers; set this option to use dedicated work queue instead.
+
+if VCMP_IT51XXX_WORKQUEUE
+
+config VCMP_IT51XXX_WORKQUEUE_PRIORITY
+	int "ITE it51xxx voltage comparator threshold detection work queue priority"
+	default SYSTEM_WORKQUEUE_PRIORITY
+	help
+	  This option sets internal ITE voltage comparator threshold detection
+	  workqueue priority.
+
+config VCMP_IT51XXX_WORKQUEUE_STACK_SIZE
+	int "ITE it51xxx voltage comparator threshold detection work queue stack size"
+	default 768
+	help
+	  This option sets internal ITE voltage comparator threshold detection
+	  workqueue stack size.
+
+endif # VCMP_IT51XXX_WORKQUEUE
```
The system workqueue is not neccesary for this driver, especially given we are working with latency sensitive hardware, which has an interrupt available. When an interrupt occurs, handle it and call the callback if set, if a user wants to defer handling of the interrupt to the system workqueue, the user can do so from the trigger callback, users who wish to handle it immediately within ISR context can do so as well :) This greatly simplifies the driver :)

Yunshao-Chiang replied:
Thanks for your recommendation.
workqueue does cause the ISR takes much time to execute (e.g. k_work_submit_to_queue() and k_work_submit() )
We should let the user decide whether workqueue is needed in callback.

bjarki-andreasen commented on drivers/comparator/comparator_it51xxx_vcmp.c:
```diff
@@ -0,0 +1,451 @@
+
+static K_KERNEL_STACK_DEFINE(vcmp_it51xxx_work_q_stack, CONFIG_VCMP_IT51XXX_WORKQUEUE_STACK_SIZE);
+
+static int vcmp_it51xxx_init_work_q(void)
+{
+	struct k_work_queue_config cfg = {
+		.name = "vcmp_work",
+		.no_yield = false,
+	};
+
+	k_work_queue_start(&vcmp_it51xxx_work_q, vcmp_it51xxx_work_q_stack,
+			   K_KERNEL_STACK_SIZEOF(vcmp_it51xxx_work_q_stack),
+			   CONFIG_VCMP_IT51XXX_WORKQUEUE_PRIORITY, &cfg);
+
+	return 0;
+}
```
Work queues are the most expensive primitive we have in zephyr, to the point where we have to share them to justify them (see system workqueue), if a driver needs one, use the system work queue, however, this driver does not need it, all execution can be done from ISR context.

bjarki-andreasen commented on drivers/comparator/comparator_it51xxx_vcmp.c:
```diff
@@ -0,0 +1,451 @@
+	/* Data must keep device reference for worker handler */
+	data->vcmp = dev;
+
+	/* Init and set work item to enable notifications */
+	k_work_init(&data->work, it51xxx_vcmp_trigger_work_handler);
+	vcmp_work_addr[cfg->vcmp_ch] = (uint32_t)&data->work;
```
use ISR directly

bjarki-andreasen commented on drivers/comparator/comparator_it51xxx_vcmp.c:
```diff
@@ -0,0 +1,451 @@
+	 * so if the irq is enabled before, we needn't to enable again.
+	 * And we will figure out the triggered channel in vcmp_it51xxx_isr().
+	 */
+	if (!irq_is_enabled(cfg->irq)) {
+		ite_intc_isr_clear(cfg->irq);
+		irq_connect_dynamic(cfg->irq, 0, (void (*)(const void *))vcmp_it51xxx_isr,
+				    (const void *)dev, 0);
+		irq_enable(cfg->irq);
+	}
```
Will this not result in only one irq handler for one comparator be connected? `irq_connect_dynamic` must be called for all of them right? with only one irq_enable(cfg->irq)?

In any case, calling irq_enable(cfg->irq) multiple times is safe, so for simplicity I would just call it every time :)

bjarki-andreasen commented on drivers/comparator/Kconfig.it51xxx_vcmp:
```diff
@@ -0,0 +1,48 @@
+config COMPARATOR_IT51XXX_VCMP
+	bool "ITE it51xxx Voltage Comparator"
+	default y
+	depends on ADC_ITE_IT51XXX
```
I believe this driver also depends on DYNAMIC_INTERRUPTS

Dino-Li commented:
@MaureenHelm @bjarki-andreasen any comments will be appreciated.

bjarki-andreasen commented:
You will need to implement is_pending, the only difference between set_trigger_callback and is_pending is that you store a flag on interrupt rather than call a callback, so it should be trivial. Once that is done, could you add an entry for this comparator to https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/comparator/gpio_loopback and test it out? the get_output case will fail for now, I will need to add support for it not being implemented, but the trigger test cases should succeed :)

Yunshao-Chiang commented:
Hi @bjarki-andreasen 

Thank you for your detailed explanation of the difference between set_trigger_callback and is_pending — it really helped clarify the implementation approach. After gaining a better understanding of how is_pending should behave, I’ve now added the function. As you mentioned, it was relatively straightforward since it only involves storing a flag on interrupt instead of invoking a callback.

I added an entry for this comparator to the [Zephyr gpio_loopback test suite](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/comparator/gpio_loopback) and run the relevant tests. I understand that the get_output case is expected to fail for now due to missing support, but the trigger-related test cases pass as expected.

Thanks again for your guidance and support!

bjarki-andreasen commented on boards/ite/it515xx_evb/it515xx_evb.dts:
```diff
@@ -153,3 +153,11 @@
 	status = "okay";
 };
+
+/* voltage comparator for test */
```
Enabling hardware for a specific test case should be done in the board overlay of the testcase :)

Yunshao-Chiang replied:
Hi @bjarki-andreasen,
Would it be okay if I delete the comment but keep the rest as user configuration reference?
I'm not sure if you think this would be appropriate.

bjarki-andreasen replied:
I really would encourage against keeping the node enabled, it will lead to users having to notice and delete it in their overlays if they want to reconfigure it. But, it is isolated to this board so I will not block it :) Keep the comment if you keep the node :)

Yunshao-Chiang replied:
@bjarki-andreasen,
I will also disable the vcmp node by default.

bjarki-andreasen reviewed (commented):
Nice work!
